in-progress/7346-batch-proving-circuits-and-l1.md

```diff
@@ -0,0 +1,259 @@
+}
+```
+
+This circuit then performs the following checks and computations:
+
+- Recursively verifies the left and right inputs
+- Checks that `right.previous_archive` equals `left.new_archive`
+- Checks that `right.previous_block_hash` equals `left.end_block_hash`
+- Checks that `right.start_global_variables` follow from `left.end_global_variables`
+- Concatenates and outputs the `fees` from both inputs
+- Outputs `sha256(left.out_hash, right.out_hash)` as its own `out_hash`
+- Outputs `previous_archive`, `start_global_variables`, and `previous_block_hash` from `left`
+- Outputs `new_archive`, `end_global_variables`, and `end_block_hash` from `right`
+
+Note that we say that the global variables in `right` "follow" the ones in `left` if:
+
+- `left.chain_id == right.chain_id`
+- `left.version == right.version`
+- `left.block_number == right.block_number + 1`
+- `left.timestamp < right.timestamp`
+- `coinbase`, `fee_recipient`, and `gas_fees` are not constrained (though `gas_fees` may be in a 1559-like world)
+
+### Root rollup
+
+The new root rollup circuit then takes two `BlockRootOrBlockMergePublicInputs`, performs the same checks as the block merge rollup, but outputs a subset of the public inputs, to make L1 verification cheaper:
+
+```rust
+struct RootRollupPublicInputs {
+  previous_archive: Field,
+  end_archive: Field,
+  end_block_hash: Field,
+  end_timestamp: Field,
+  end_block_number: Field,
+  out_hash: Field,
+  fees: [{ recipient: Address, value: Field }; 32]
+}
+```
+
+### Empty block root rollup
+
+Since we no longer submit a proof per block, and thanks to @MirandaWood we now have wonky rollups, we no longer need to fill a block with "empty" txs. This means we can discard the empty private kernel circuit.
+
+However, we still need to be able to represent an empty block. We can do this by introducing an empty block root rollup circuit, which outputs a `BlockRootOrBlockMergePublicInputs` which does not consume any merge rollups, and the end state just equals the start state. Note that we may still need an empty nested circuit to fill in the nested proof.
+
+## L1 Rollup Contract
+
+The Rollup contract today has a main entrypoint `process(header, archive, aggregationObject, proof)`. We propose breaking this method into two:
+
+```solidity
+contract Rollup {
+  process(header, archive);
+  submitProof(publicInputs, aggregationObject, proof);
+}
+```
+
+### State
+
+To track both the proven and unproven chains, we add the following state variables to the contract:
+
+```diff
+contract Rollup {
+  bytes32 lastArchiveTreeRoot;
+  uint256 lastBlockTimestamp;
++ bytes32 verifiedArchiveTreeRoot;
++ uint256 verifiedBlockTimestamp;
+}
+```
+
+The `last*` fields are updated every time a new block is uploaded, while the `verified*` ones are updated when a proof is uploaded. In the event of a rollback due to failure on proof submission, the `last*` fields are overwritten with the contents of `verified*`.
```

Comment:
Maybe we need another set of variables to track in-flight proving, otherwise publishing the 33rd block would overwrite the `last*` variables in the rollup which would fail proof validation.

Probably something to think about for proving coordination and having just a verified/last set of variables is enough.

Reply:
> otherwise publishing the 33rd block would overwrite the last* variables in the rollup which would fail proof validation

Agree! That's why I suggested moving the "prove that this block root is in the current archive tree" to Solidity instead of the root rollup circuit. The proof would have a "last block proven", and the submitter would include a merkle path from it to the current `lastArchiveTreeRoot`. If that root changes, then the submitter needs to generate a new merkle path instead of rebuilding the entire root proof.

To further mitigate against this race condition, we could keep the last N archive tree roots instead of just the latest.

Reply:
A fairly simple approach that I think I would take for the first version at least, would be to simple store a "list" of the archive values instead of just the tips. 

This could be as:
```solidity
mapping(uint256 blockNumber => bytes32 archive) public archives;
```

Then you keep track of 2 indexes for how far we have added blocks, and how far we have proven.

This would also allow providing a proof that is not the "next", e.g., to avoid the issue where someone is delaying their proof to limit the time you have to interact. We are looking to add some of this in https://github.com/AztecProtocol/aztec-packages/issues/7614.

Reply:
@LHerskind sounds good as a first approach, given that L1 gas is free!